test/profile.spec.js

```diff
@@ -0,0 +1,67 @@
+'use strict'
+
+const t = require('tap')
+const { readFileSync } = require('fs')
+const { file: tmpFile } = require('tmp-promise')
+
+const generateHeapSamplingProfile = require('../src/profile')
+const allocateMemoryFor = require('./allocator')
+const validate = require('./fixtures/profileSchema')
+
+t.test('it correctly generates a profile using promises', async t => {
+  const { path: destination, cleanup } = await tmpFile()
+
+  await Promise.all([generateHeapSamplingProfile({ destination, duration: 100 }), allocateMemoryFor(100)])
+
+  const generated = JSON.parse(readFileSync(destination, 'utf-8'))
+
+  t.true(validate(generated))
+
+  cleanup()
+})
+
+t.test('it correctly generates a profile using callbacks', t => {
+  Promise.all([
+    allocateMemoryFor(100),
+    tmpFile().then(({ path: destination, cleanup }) => {
+      generateHeapSamplingProfile({ destination, duration: 100 }, () => {
```

Comment:
mixing callbacks and promises is really ugly and can have bugs. Can you please detangle this?

Reply:
Agree. Changed to a "nicer" version.